importers/src/bible/osis/index.ts

```diff
@@ -0,0 +1,945 @@
+                // ignore (until needed)
+                break;
+            default: {
+                if (DEBUG_OUTPUT_ENABLED) console.log(`unrecognized osis xml tag: ${elementType}`);
+            }
+        }
+    }
+
+    parseClosingTag(tagName: OsisXmlNodeName, context: ParserContext) {
+        if (context.skipClosingTags.length && context.skipClosingTags[0] === tagName) {
+            context.skipClosingTags.shift();
+            if (context.openedSelfClosingTag) delete context.openedSelfClosingTag;
+            return;
+        }
+
+        let currentTag: ITagWithType | undefined;
+        if (context.openedSelfClosingTag) {
+            if (context.openedSelfClosingTag.name !== tagName) {
+                throw new Error(`invalid self closing end tag`);
+            }
+            currentTag = context.openedSelfClosingTag;
+            delete context.openedSelfClosingTag;
+            // stop if this is actually a starting tag
+            if (currentTag.attributes.sID) return;
+        } else {
+            currentTag = context.hierarchicalTagStack.pop();
+        }
+
+        if (!currentTag) throw new Error(`can't find matching tag for closing tag ${tagName}`);
+
+        let closeTagsAtEnd: OsisXmlNodeName[] = [];
+        let startTagsAtEnd: OsisXmlNode[] = [];
+
+        if (currentTag.name !== tagName) {
+            // an unclosed quote group?
+            // RADAR: since some quote span over multiple chapters, opening a quote group leads to a
+            //     corrupted document structure in a lot of cases. The following code blocks try to
+            //     fix this automatically. However since this doesn't work reliably, we currently
+            //     only create phrases for the quote markers, not groups.
+            //     Since having proper quote groups would be preferred (to retain the
+            //     semantic information in BibleEngine) we leave this code here
+            //     in case this can be implemented properly at a later stage
+            if (currentTag.name === OsisXmlNodeName.QUOTE) {
+                const quoteContainer = this.getCurrentContainer(context);
+                if (
+                    !quoteContainer ||
+                    quoteContainer.type !== 'group' ||
+                    quoteContainer.groupType !== 'quote'
+                )
+                    throw new Error(
+                        this.getErrorMessageWithContext(
+                            `expected quote group on top of stack`,
+                            context
+                        )
+                    );
+
+                const outerTag = context.hierarchicalTagStack.pop();
+                if (!outerTag || outerTag.name !== tagName)
+                    throw new Error(`invalid closing tag: ${tagName} (after second level)`);
+
+                const outerContainer =
+                    context.contentContainerStack[context.contentContainerStack.length - 2];
```

Comment:
is this safe? makes me nervous, haha

Reply:
yes, it just returns undefined if the index is not there. since we are not accessing the object directly this is safe.